tchannel/tornado/connection.py

```diff
@@ -358,8 +365,15 @@ def _write(self, message):
             payload=payload
         )
         body = frame.frame_rw.write(f, BytesIO()).getvalue()
+        return self._write_queue.put(body)
 
-        return self.connection.write(body)
+    @tornado.gen.coroutine
```

Comment:
You can drop the `gen.coroutine` by making this future-callback based.

```python
def _drain_write_queue(self):
    io_loop = IOLoop.current()

    def _on_write(future):
        io_loop.spawn_callback(_write_next)
        if future.exception():    
            log(future.exception())

    def _on_message(future):
        if future.exception():
            io_loop.spawn_callback(_write_next)
            log(future.exception())
            return
        message = future.result()
        io_loop.add_future(self.connection.write(message), _on_write)

    def _write_next():
        if self.closed:
            return
        io_loop.add_future(self._write_queue.get(), _on_message)

    io_loop.spawn_callback(_write_next)
```

Note that the use of `spawn_callback` makes it drop any stack context knowledge during the write. That should reduce overhead. If we want to keep that information, we'll have to use `add_callback`.